kernel/src/syscall/prlimit64.rs

```diff
@@ -51,59 +33,75 @@ pub fn sys_prlimit64(
-
-        if old_rlim_addr != 0 {
-            let rlimit = resource_limits.get_rlimit(resource);
-            let (cur, max) = rlimit.get_cur_and_max();
-            let rlimit_raw = RawRLimit64 { cur, max };
-            ctx.user_space().write_val(old_rlim_addr, &rlimit_raw)?;
-        }
-        if new_rlim_addr != 0 {
-            let new_raw: RawRLimit64 = ctx.user_space().read_val(new_rlim_addr)?;
-            debug!("new_rlimit = {:?}", new_raw);
-            if resource == ResourceType::RLIMIT_NOFILE && new_raw.max > SYSCTL_NR_OPEN {
-                return_errno_with_message!(Errno::EPERM, "the new limit exceeds the system limit");
-            }
-
-            resource_limits
-                .get_rlimit(resource)
-                .set_cur_and_max(new_raw.cur, new_raw.max)?;
-        }
-
-        Ok(())
+    let new_raw = if new_rlim_addr == 0 {
+        None
+    } else {
+        Some(userspace.read_val(new_rlim_addr)?)
     };
 
-    if pid == 0 || pid == ctx.process.pid() {
-        get_and_set_rlimit(ctx.process.as_ref())?;
+    let old_raw = if pid == 0 || pid == ctx.process.pid() {
+        do_prlimit64(&ctx.process, resource, new_raw, ctx)?
     } else {
         let target_process = process_table::get_process(pid).ok_or_else(|| {
             Error::with_message(Errno::ESRCH, "the target process does not exist")
         })?;
-
         // Check permissions
-        check_rlimit_perm(target_process.main_thread().as_posix_thread().unwrap(), ctx)?;
-        get_and_set_rlimit(target_process.as_ref())?;
+        check_rlimit_perm(&target_process, ctx)?;
+        do_prlimit64(&target_process, resource, new_raw, ctx)?
+    };
+
+    if old_rlim_addr != 0 {
+        userspace.write_val(old_rlim_addr, &old_raw)?;
     }
+
     Ok(SyscallReturn::Return(0))
 }
 
+fn do_prlimit64(
+    target_process: &Process,
+    resource: u32,
+    new_raw: Option<RawRLimit64>,
+    ctx: &Context,
+) -> Result<RawRLimit64> {
+    let resource = ResourceType::try_from(resource)?;
+    debug!(
+        "pid = {}, resource = {:?}, new_raw = {:?}",
+        target_process.pid(),
+        resource,
+        new_raw,
+    );
+
+    let rlimit = {
+        let resource_limits = target_process.resource_limits();
+        resource_limits.get_rlimit(resource)
+    };
+
+    let old_raw = {
+        let (cur, max) = rlimit.get_cur_and_max();
+        RawRLimit64 { cur, max }
+    };
+
+    if let Some(new_raw) = new_raw {
+        debug!("new_rlimit = {:?}", new_raw);
+        if resource == ResourceType::RLIMIT_NOFILE && new_raw.max > SYSCTL_NR_OPEN {
+            return_errno_with_message!(Errno::EPERM, "the new limit exceeds the system limit");
+        }
+        if new_raw.max > old_raw.max {
+            let init_user_ns = UserNamespace::get_init_singleton();
+            init_user_ns.check_cap(CapSet::SYS_RESOURCE, ctx.posix_thread)?;
+        }
+        rlimit.set_cur_and_max(new_raw.cur, new_raw.max)?;
```

Comment:
`get_cur_and_max` & `set_cur_and_max` should be an atomic operation? I guess it's an existing problem, though.
https://elixir.bootlin.com/linux/v6.18.2/source/kernel/sys.c#L1519-L1538